module/zfs/arc.c

```diff
@@ -8018,9 +8018,27 @@ l2arc_write_done(zio_t *zio)
 	list_destroy(&cb->l2wcb_abd_list);
 
 	if (zio->io_error != 0) {
-		/* restore the lbps array in the header to its previous state */
+		/*
+		 * Restore the lbps array in the header to its previous state.
+		 * If the list of log block pointers is empty, zero out the
+		 * log block pointers in the device header.
+		 */
 		lb_ptr_buf = list_head(&dev->l2ad_lbptr_list);
 		for (int i = 0; i < 2; i++) {
+			if (lb_ptr_buf == NULL) {
+				/*
+				 * If the list is empty zero out the device
+				 * header. Otherwise zero out the second log
+				 * block pointer in the header.
+				 */
+				if (i == 0) {
+					bzero(l2dhdr, dev->l2ad_dev_hdr_asize);
+				} else {
+					bzero(&l2dhdr->dh_start_lbps[i],
+					    sizeof (l2arc_log_blkptr_t));
+				}
+				break;
+			}
```

Comment:
I'm confused by this:
```
		for (int i = 0; i < 2; i++) {
			if (lb_ptr_buf == NULL) {
				/*
				 * If the list is empty zero out the device
				 * header. Otherwise zero out the second log
				 * block pointer in the header.
				 */
				if (i == 0) {
					bzero(l2dhdr, dev->l2ad_dev_hdr_asize);
				} else {
					bzero(&l2dhdr->dh_start_lbps[i],
					    sizeof (l2arc_log_blkptr_t));
				}
				break;
			}
```
It says if the list is empty zero out the device, otherwise (meaning if the list is not empty?) zero out the second log.

What I'm reading from the code is that if the list is empty first zero out the device header, then zero out `&l2dhdr->dh_start_lbps[i]`.

Reply:
I will clarify the comment in the code. The device header has two pointers to the two most recently written log blocks (L2ARC metadata). The for loop serves restore them in their previous state, as the current zio failed (and thus did not write any log blocks).

We enter the for loop. If `lb_ptr_buf` is NULL --> this means the `l2ad_lbptr_list` is empty. So we zero out the whole device header, including both log block pointers, as there are no valid log blocks.

However if `l2ad_lbptr_list` contains only one log block pointer, then the first entry l2dhdr->dh_start_lbps[0] is set to that.
Then the for loop iterates, so now i=1. Since `l2ad_lbptr_list` contains only one entry, `l2ad_ptr_buf` is going to be NULL this time around. Then only the second log block pointer in the device header l2ad_lbptr_list[1] is going to be zeroed out.
Does this make sense?

I will improve that comment.

Reply:
Ah! That makes sense. I completely missed the `list_next` at the end of the loop and got confused. Marking as resolved.